p2p/router.go

```diff
@@ -0,0 +1,557 @@
+// time, the peerStore will also do peer scheduling and prioritization, e.g.
+// ensuring we do exponential backoff on dial failures and connecting to
+// more important peers first (such as persistent peers and validators).
+//
+// An additional goroutine Router.broadcastPeerUpdates() is also spawned off
+// on startup, which consumes peer updates from Router.peerUpdatesCh (currently
+// only connections and disconnections), and broadcasts them to all peer update
+// subscriptions registered via SubscribePeerUpdates().
+//
+// On router shutdown, we close Router.stopCh which will signal to all
+// goroutines to terminate. This in turn will cause all pending channel/peer
+// queues to close, and we wait for this as a signal that goroutines have ended.
+//
+// All message scheduling should be limited to the queue implementations used
+// for channel queues and peer queues. All message sending throughout the router
+// is blocking, and if any messages should be dropped or buffered this is the
+// sole responsibility of the queue, such that we can limit this logic to a
+// single place. There is currently only a FIFO queue implementation that always
+// blocks and never drops messages, but this must be improved with other
+// implementations. The only exception is that all message sending must also
+// select on appropriate channel/queue/router closure signals, to avoid blocking
+// forever on a channel that has no consumer.
+type Router struct {
+	*service.BaseService
+	logger     log.Logger
+	transports map[Protocol]Transport
+	store      *peerStore
+
+	// FIXME: Consider using sync.Map.
+	peerMtx    sync.RWMutex
+	peerQueues map[NodeID]queue
+
+	// FIXME: We don't strictly need to use a mutex for this if we seal the
+	// channels on router start. This depends on whether we want to allow
+	// dynamic channels in the future.
+	channelMtx      sync.RWMutex
+	channelQueues   map[ChannelID]queue
+	channelMessages map[ChannelID]proto.Message
+
+	peerUpdatesCh   chan PeerUpdate
+	peerUpdatesMtx  sync.RWMutex
+	peerUpdatesSubs map[*PeerUpdatesCh]*PeerUpdatesCh // keyed by struct identity (address)
+
+	// stopCh is used to signal router shutdown, by closing the channel.
+	stopCh chan struct{}
+}
+
+// NewRouter creates a new Router, dialing the given peers.
+//
+// FIXME: providing protocol/transport maps is cumbersome in tests, we should
+// consider adding Protocols() to the Transport interface instead and register
+// protocol/transport mappings automatically on a first-come basis.
+func NewRouter(logger log.Logger, transports map[Protocol]Transport, peers []PeerAddress) *Router {
+	router := &Router{
+		logger:          logger,
+		transports:      transports,
+		store:           newPeerStore(),
+		stopCh:          make(chan struct{}),
+		channelQueues:   map[ChannelID]queue{},
+		channelMessages: map[ChannelID]proto.Message{},
+		peerQueues:      map[NodeID]queue{},
+		peerUpdatesCh:   make(chan PeerUpdate, 1),
```

Comment:
Just in case the receiving goroutine is busy broadcasting the previous update we don't hold up the connection. But this should be benchmarked, will remove it for now.